sympy/core/function.py

```diff
@@ -909,6 +922,7 @@ def __new__(mcl, name, bases=(AppliedUndef,), __dict__=None, **kwargs):
         __dict__['__module__'] = None
         obj = super(UndefinedFunction, mcl).__new__(mcl, name, bases, __dict__)
         obj.name = name
+        obj._sage_ = UndefSage()
```

Comment:
Does this need to be a different instance for each undefined function? As far as I can tell the UndefSage class has no instance data so it can be a singleton class. In fact why isn't it just a property-method using `@property`?

Reply:
I changed it to a Singleton class.

>  In fact why isn't it just a property-method using @property?

What do you mean by this?

Reply:
Can we instead define a method like this?
```python
    @property
    def _sage_(self):
        ...
```
It isn't normally necessary to explicitly create a descriptor object.

Reply:
No, this is a class with `AppliedUndef` as a base and therefore its `_sage_` is called instead of the method like you mentioned.

Reply:
I meant that it's a meta class.

Reply:
Why can't the `_sage_` method in `AppliedUndef` do what this does?

I don't mean to be difficult but all this metaclass stuff is awkward and I don't want to make it any more complicated than it needs to be (longer term I want to remove the metaclasses here).

Reply:
Because that's for the instance of `AppliedUndef`. This method is for the class itself. See also the PR description for why it was done this way.

Reply:
The PR description is inaccurate. If I delete the line that we are discussing then:
```julia
In [5]: F = Function('F')                                                                                                                      

In [6]: F._sage_                                                                                                                               
Out[6]: <function sympy.core.function.AppliedUndef._sage_(self)>
```
So `F._sage_` does indeed refer to the `AppliedUndef` method.

Reply:
That's the issue, isn't it? You can't call `F._sage_()` and you can only call `F(x)._sage_()`

Reply:
It does seem that way (I'm trying to think of an alternative...)

Reply:
Btw when I said the class could be a singleton I just meant that it could be like:
```python
class A:
    pass
a = A()
...
obj._sage_ = a
```
rather than using more metaclasses.

Reply:
Wouldn't that increase import time for sympy?

Reply:
I don't see why it would increase import time. Here's a quick test (module `t.py`):
```python
from sympy.core.compatibility import with_metaclass
from sympy.core.singleton import Singleton

def f():
    class A(with_metaclass(Singleton)):
        pass

def g():
    class B:
        pass
    b = B()
```
Running this with timeit I get:
```console
$ python -m timeit -s 'from t import f, g' 'f()'
1000 loops, best of 5: 222 usec per loop
$ python -m timeit -s 'from t import f, g' 'g()'
50000 loops, best of 5: 9.72 usec per loop
```
So normal Python classes are 20 times faster than Singleton/metaclass.

The purpose of the Singleton metaclass is to install the class in the singleton registry so it becomes available as e.g. `S.Infinity` which isn't what we want here.

Reply:
Thanks @oscarbenjamin, I've removed the use of `Singleton` metaclass and used the method you proposed.